### PR TITLE
Alliance Chat Mark I

### PR DESCRIPTION
**Pull request by daprof1**
Long awaited
Often Imitated
Never Duplicated

Alliance Chat is here, I'm sure some of the kludgy code could be cleaned up, take a look and lemme know.

### CODE REVIEW

File: airline-web/app/controllers/ChatApplication.scala [hunk 1]
```diff
@@ -17,9 +21,19 @@ class ChatApplication @Inject()(actorSystem: ActorSystem) extends Controller {
...
-    (out: ActorRef) =>
-    Props(new ClientActor(out, chat))
-  }	 
+ // def chatSocket = WebSocket.acceptWithActor[String, String] {
+ //   (request: RequestHeader) =>
+ //   (out: ActorRef) =>
+ //   Props(new ClientActor(out, chat))
+ //  }	 
+   def chatSocket = WebSocket.tryAcceptWithActor[String, String] { request => 
+    Future.successful(request.session.get("userId") match {
```
Comment (patsonluk) [inline airline-web/app/controllers/ChatApplication.scala]: Wondering where does the userId come from? Is it authenticated :) ?

File: airline-web/app/controllers/ClientActor.scala [hunk 2]
```diff
@@ -18,10 +25,38 @@ class ClientActor(out: ActorRef, chat: ActorRef) extends Actor {
...
+		  val airline = AirlineSource.loadAirlineById(userId)
+		  val otext =  "[" + sdf.format(Calendar.getInstance().getTime())+ "] " + airline.get.name + ": " + json_text.\("text").as[String]
+		  val room = json_text.\("room").as[String]
+		  var o_room = -1
+		  if (room != "chatBox-1") {
+			// Get Alliance ID
+			o_room = 0; // Default
+			val airlines = Map[Int, Airline]()
+			val airline2 = airlines.getOrElseUpdate(userId, AirlineSource.loadAirlineById(userId, false).getOrElse(Airline.fromId(userId)))
+			  AllianceSource.loadAllianceMemberByAirline(airline2).foreach { member =>
```
Comment (patsonluk) [inline airline-web/app/controllers/ClientActor.scala]: Why do we need to reload the airline here? what's the map `airlines` for? The airlines.getOrElseUpdate will always run the update part as the Map is empty from the previous statement?

File: airline-web/app/models/Chat.scala [hunk 3]
```diff
@@ -12,7 +14,12 @@ final case class ClientSentMessage(text: String)
 // Chat actor
 class Chat extends Actor {
   // initial message-handling behavior
-  var log = Queue[String]()
+  
+  // General Chat Log (50)
+  var g_log = Queue[String]()
```
Comment (patsonluk) [inline airline-web/app/models/Chat.scala]: should be `val`?

File: airline-web/app/controllers/ChatApplication.scala (hunk 1 above)
Comment (daprof1) [reply airline-web/app/controllers/ChatApplication.scala]: I copied that from your code .. you tell me! But it's the authenticated ID

File: airline-web/app/controllers/ClientActor.scala (hunk 2 above)
Comment (daprof1) [reply airline-web/app/controllers/ClientActor.scala]: It was literally the only way I could figure out how to pull the alliance id if it exists .. I'm all for a better way but I wasn't gonna touch the model .. I'll let you put in an getAllianceIDbyAirlineID

File: airline-web/app/models/Chat.scala (hunk 3 above)
Comment (daprof1) [reply airline-web/app/models/Chat.scala]: don't think so as it's constantly modified, might as well keep it mutable.

File: airline-web/app/controllers/ChatApplication.scala (hunk 1 above)
Comment (patsonluk) [reply airline-web/app/controllers/ChatApplication.scala]: ah ok it's from `AuthenticationObject` 😅

File: airline-web/app/controllers/ClientActor.scala (hunk 2 above)
Comment (patsonluk) [reply airline-web/app/controllers/ClientActor.scala]: cant u just replace airline2 with airline.get?

File: airline-web/app/models/Chat.scala (hunk 3 above)
Comment (patsonluk) [reply airline-web/app/models/Chat.scala]: the Queue is mutable but not the reference i suppose?

File: airline-web/app/controllers/ClientActor.scala (hunk 2 above)
Comment (daprof1) [reply airline-web/app/controllers/ClientActor.scala]: I'll try it again

Comment (daprof1) [reply airline-web/app/controllers/ClientActor.scala]: Okay redone

File: airline-web/app/models/Chat.scala (hunk 3 above)
Comment (daprof1) [reply airline-web/app/models/Chat.scala]: changed it

File: airline-web/app/controllers/ClientActor.scala [hunk 4]
```diff
@@ -18,10 +25,37 @@ class ClientActor(out: ActorRef, chat: ActorRef) extends Actor {
       if (text.indexOf("[LOGGED]") > -1) {
 	   chat ! ClientSentMessage(text)
 	  } else {
-	  chat ! ClientSentMessage("[" + sdf.format(Calendar.getInstance().getTime())+ "] " + text)
+		  val json_text: JsValue = Json.parse(text)
+		  val airline = AirlineSource.loadAirlineById(userId)
```
Comment (patsonluk) [inline airline-web/app/controllers/ClientActor.scala]: Merged the code in but this line is not quite right...airlineId is not the same as userId, it's throwing some exceptions in prod now

Not sure if we want to load the airline again here. Do we really have to?

**Pull request merged**